### PR TITLE
topdown: Fixing issues with optimizing rules with refs in the head

### PR DESCRIPTION
**Pull request by johanfylling**
Issues fixed:

* production of support module with forbidden characters in first var of rule ref (#6338)
* panic when policy contains rules with a general ref in the head (#6339)

### CODE REVIEW

File: ast/term.go [hunk 1]
```diff
@@ -1102,7 +1102,7 @@ func (ref Ref) Ptr() (string, error) {
 	return strings.Join(parts, "/"), nil
 }
 
-var varRegexp = regexp.MustCompile("^[[:alpha:]_][[:alpha:][:digit:]_]*$")
+var VarRegexp = regexp.MustCompile("^[[:alpha:]_][[:alpha:][:digit:]_]*$")
```
Comment (srenatus) [inline ast/term.go:1105]: I think I'd rather avoid exporting the RE and do something like
```rego
func IsVar(s string) bool {
    return varRegexp.MatchString(s)
}
```
instead. Would it do the trick, here, too?

Comment (johanfylling) [reply ast/term.go:1105]: Good point 👍 . I'll fix that.

File: topdown/save.go [hunk 2]
```diff
@@ -317,6 +332,25 @@ func (s *saveSupport) Insert(path ast.Ref, rule *ast.Rule) {
 	module.Rules = append(module.Rules, rule)
 }
 
+func splitPackageAndRule(path ast.Ref) (ast.Ref, ast.Ref) {
+	p := path.Copy()
+
+	ruleRefStart := 2 // path always contains at least 3 terms (data. + one term in package + rule name)
```
Comment (srenatus) [inline topdown/save.go:338]: 💭 Until we make `package` optional 😉 (Kidding, it's OK as-is.)

File: topdown/save.go [hunk 3]
```diff
@@ -317,6 +332,25 @@ func (s *saveSupport) Insert(path ast.Ref, rule *ast.Rule) {
...
+	for i := ruleRefStart; i < len(p.StringPrefix()); i++ {
+		t := p[i]
+		if str, ok := t.Value.(ast.String); ok && ast.IsVarCompatibleString(string(str)) {
```
Comment (srenatus) [inline topdown/save.go:341]: 💭 We're turning an `ast.String` into a string to feed it into `ast.IsVarCompatibleString`. Perhaps the latter should just take an `ast.String`? They're both "ast things"...

File: topdown/eval.go [hunk 4]
```diff
@@ -2694,35 +2694,31 @@ func (e evalVirtualPartial) partialEvalSupportRule(rule *ast.Rule, path ast.Ref)
...
+			if rule.Head.Key != nil && rule.Head.RuleKind() == ast.MultiValue {
+				head.Key = child.bindings.PlugNamespaced(rule.Head.Key, e.e.caller.bindings)
+			}
+
+			if rule.Head.RuleKind() == ast.SingleValue && len(ruleRef) == 2 {
+				head.Key = ruleRef[len(ruleRef)-1]
+			}
+
+			if head.Name.Equal(ast.Var("")) && (len(ruleRef) == 1 || (len(ruleRef) == 2 && rule.Head.RuleKind() == ast.SingleValue)) {
+				head.Name = ruleRef[0].Value.(ast.Var)
```
Comment (srenatus) [inline topdown/eval.go:2721]: So we're basically back-filling `Name` now for ref-rules?
```rego
p.q.r[x][y][z] := 1
```
would have name `p`?

**Review by srenatus — COMMENTED** [review]
Couple of comments and questions, but it looks good -- it's passing the tests! 😅

File: topdown/eval.go (hunk 4 above)
Comment (johanfylling) [reply topdown/eval.go:2721]: Only if the length is max 2. I.e. an old style "non-ref" partial object or set

**Review by srenatus — APPROVED** [review]

**Pull request merged**